### PR TITLE
Introduce HTML coverage report

### PR DESCRIPTION
**Pull request by Morriar**
This PR reuse all the features from snapshot, timeline and coverage to create a beautiful report.

Here's how to generate the report for `spoom`

```
cd spoom
spoom coverage timeline --save
spoom coverage report
spoom coverage open
```

And here's the output:

image
image
image
image

### CODE REVIEW

**Comment by Morriar** [comment]
Also the map is interactive 😄

**Review by paracycle — APPROVED** [review]
This is amazing progress. 🎉  However, I think we should evolve this so that the Ruby generation part and the HTML/JS/CSS part have a well defined interface for what and how data is going to be passed between them. If we do that, we should be able to edit each component independently and open the door for users to replace templates with their own things, if they want.

**Comment by Morriar** [comment]
New version using Erb templates for the main HTML components.

I didn't use `Thor::actions` as to not generate files all the time so the report can be inherited/manipulated before output from clients.

File: lib/spoom/cli/commands/coverage.rb [hunk 1]
```diff
@@ -103,6 +103,47 @@ def timeline
...
+          files = Dir.glob("#{data_dir}/*.json")
+          if files.empty?
+            message_no_data(data_dir)
+            exit(1)
+          end
+
+          snapshots = files.sort.map do |file|
+            json = File.read(file)
+            Spoom::Coverage::Snapshot.from_json(json)
+          end.filter(&:commit_timestamp).sort_by!(&:commit_timestamp)
```
Comment (KaanOzkan) [inline lib/spoom/cli/commands/coverage.rb:122]: We actually have items/commits without a timestamp here? 😲

File: lib/spoom/config.rb [hunk 2]
```diff
@@ -6,5 +6,6 @@ module Config
     SORBET_CONFIG = "sorbet/config"
     SORBET_GEM_PATH = Gem::Specification.find_by_name("sorbet-static").full_gem_path
     SORBET_PATH = (Pathname.new(SORBET_GEM_PATH) / "libexec" / "sorbet").to_s
+    SPOOM_PATH = (Pathname.new(__FILE__) / ".." / ".." / "..").to_s
```
Comment (KaanOzkan) [inline lib/spoom/config.rb:9]: I'm curious what's going on here. I think you are moving back to the root directory of this repo but i don't know how. 

What does the `/` do?

File: lib/spoom/coverage.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    sig { params(path: String).returns(Snapshot) }
+    def self.snapshot(path: '.')
+      snapshot = Snapshot.new
+      metrics = Spoom::Sorbet.srb_metrics(path: path, capture_err: true)
+      return snapshot unless metrics
+
+      sha = Spoom::Git.last_commit(path: path)
+      snapshot.commit_sha = sha
+      snapshot.commit_timestamp = Spoom::Git.commit_timestamp(sha, path: path).to_i if sha
```
Comment (KaanOzkan) [inline lib/spoom/coverage.rb:22]: I guess my previous comment is invalid now and the question is this `sha` can be null 😲

**Review by KaanOzkan — APPROVED** [review]
I skimmed over the CSS/D3JS parts 😕

File: lib/spoom/config.rb (hunk 2 above)
Comment (paracycle) [reply lib/spoom/config.rb:9]: `Pathname#/` is an alias to `Pathname#join` :)

**Review by paracycle — APPROVED** [review]
Some comments about removing some of the presentational things from the Template classes and moving them into the actual templates, so that the template is just build from the data in the Template object itself, no extra presentational stuff coming from Ruby code.

Otherwise, this looks great.

File: lib/spoom/cli/commands/coverage.rb (hunk 1 above)
Comment (Morriar) [reply lib/spoom/cli/commands/coverage.rb:122]: Spoom allows you to take a snapshot of a project that doesn't uses Git so no commit sha nor timestamp (and also no timeline).

**Comment by Morriar** [comment]
> Some comments about removing some of the presentational things from the Template classes and moving them into the actual templates, so that the template is just built from the data in the Template object itself, no extra presentational stuff coming from Ruby code.

I've been doing back and forth on that.

Basically what I wanted (for example in Sorbet-Metrics) was to be able to customize the header/footer and style while controlling the content (through the cards) but without rewriting the template files.

The rationale behind this version is that you can customize the default template directly from your script without having to duplicate it thanks to inheritance.

I'm happy to change it but I feel the proposed approach would force me to rewrite the whole template file just to change a few details in each client.

**Comment by paracycle** [comment]
> The rationale behind this version is that you can customize the default template directly from your script without having to duplicate it thanks to inheritance.

OK, that's fair and is a legitimate use-case.

> I'm happy to change it but I feel the proposed approach would force me to rewrite the whole template file just to change a few details in each client.

I don't feel that strongly about it but I feel like, ultimately, the best thing is to customize templates instead of code. But we can evolve that as part of later work, not a requirement for this PR to be shipped as is.

**Pull request merged**